compiler/rustc_trait_selection/src/solve/eval_ctxt/canonical.rs

```diff
@@ -81,12 +83,13 @@ impl<'tcx> EvalCtxt<'_, 'tcx> {
             }
         };
 
-        let canonical = Canonicalizer::canonicalize(
+        let mut canonical = Canonicalizer::canonicalize(
```

Comment:
as a general thought, i would expect deduplication to happen before canonicalization.

i.e. as part of `compute_external_query_constraints`.